### PR TITLE
[vcpkg] Turn on tests in CI.

### PR DESCRIPTION
**Pull request by BillyONeal**

### CODE REVIEW

**Comment by BillyONeal** [comment]
Example failure: https://dev.azure.com/vcpkg/public/_build/results?buildId=36291&view=logs&j=657c9d73-2672-59f2-b256-fdad23962c37&t=db425a18-2931-5c88-9ee2-7d283bf797fd

**Comment by BillyONeal** [comment]
I filed #11240 to track symlink tests not being able to run in our CI system.

**Comment by strega-nil** [comment]
I would prefer it if we just ignore the bootstrap scripts for this purpose; there's no reason to not just build with CMake, IMO.

File: scripts/azure-pipelines/osx/azure-pipelines.yml [hunk 1]
```diff
@@ -30,7 +30,9 @@ jobs:
     displayName: 'Build vcpkg'
     inputs:
       filePath: bootstrap-vcpkg.sh
-      arguments: '-allowAppleClang'
+      arguments: '-allowAppleClang -buildTests'
```
Comment (strega-nil) [inline scripts/azure-pipelines/osx/azure-pipelines.yml:33]: I know you didn't add this, but `-allowAppleClang` shouldn't be required.

File: scripts/azure-pipelines/windows/run-tests.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    inputs:
+      script: |
+        :: TRANSITION, get these tools on the VMs next time we roll them
+        call .\bootstrap-vcpkg.bat
+        .\vcpkg.exe fetch cmake
+        .\vcpkg.exe fetch ninja
+        set PATH=downloads\tools\cmake-3.14.0-windows\cmake-3.14.0-win32-x86\bin;downloads\tools\ninja-1.10.0-windows;%PATH%
+        call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\Common7\Tools\VsDevCmd.bat" -arch=x86 -host_arch=x86
+        rmdir /s /q build.x86.debug > nul 2> nul
+        cmake.exe -G Ninja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_BUILD_TYPE=Debug -DBUILD_TESTING=ON -DVCPKG_DEVELOPMENT_WARNINGS=ON -DVCPKG_WARNINGS_AS_ERRORS=ON -DVCPKG_BUILD_FUZZING=ON -B build.x86.debug -S toolsrc
```
Comment (strega-nil) [inline scripts/azure-pipelines/windows/run-tests.yml:22]:
```suggestion
        cmake.exe -G Ninja -DCMAKE_BUILD_TYPE=Debug -DBUILD_TESTING=ON -DVCPKG_DEVELOPMENT_WARNINGS=ON -DVCPKG_WARNINGS_AS_ERRORS=ON -DVCPKG_BUILD_FUZZING=ON -B build.x86.debug -S toolsrc
```

Also, why are you building fuzzing? just to make sure it builds?

Comment (BillyONeal) [reply scripts/azure-pipelines/windows/run-tests.yml:22]: Yes.

**Comment by BillyONeal** [comment]
> I would prefer it if we just ignore the bootstrap scripts for this purpose; there's no reason to not just build with CMake, IMO.

This is a policy question vis a vis version of cmake we work with. The version of cmake that is on the system is much older than the one bootstrap will install. Do we want to work with the version of cmake that comes with ubuntu 18.04?

**Comment by strega-nil** [comment]
Hmm, fair point. We could use pip to install an up to date version tho

**Comment by BillyONeal** [comment]
@ras0219 @ras0219-msft Assigning this to you because those test failures are preexisting and I think you're the only one from around when that code was written.

File: toolsrc/src/vcpkg/base/system.process.cpp [hunk 3]
```diff
@@ -365,16 +383,25 @@ namespace vcpkg
...
-                                                TRUE,
-                                                IDLE_PRIORITY_CLASS | CREATE_UNICODE_ENVIRONMENT | dwCreationFlags,
-                                                (void*)(env.m_env_data.empty() ? nullptr : env.m_env_data.data()),
-                                                nullptr,
-                                                &startup_info,
-                                                &process_info.proc_info);
+
+#if defined(_MSC_VER)
+#pragma warning(suppress : 6335) // Leaking process information handle 'process_info.proc_info.hProcess'
+                                 // /analyze can't tell that we transferred ownership here
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/base/system.process.cpp:389]: Would it be possible to do something like:

```c++
MSVC_SUPPRESS_WARNING(6335)
```

as opposed to this `#if/#endif` thing we do now?

File: toolsrc/src/vcpkg/base/system.process.cpp [hunk 4]
```diff
@@ -327,29 +327,47 @@ namespace vcpkg
...
     struct ProcessInfo
     {
-        constexpr ProcessInfo() : proc_info{} {}
+        constexpr ProcessInfo() noexcept : proc_info{} { }
+        ProcessInfo(ProcessInfo&& other) noexcept : proc_info(other.proc_info)
+        {
+            other.proc_info.hProcess = nullptr;
+            other.proc_info.hThread = nullptr;
+        }
+        ~ProcessInfo()
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/base/system.process.cpp:336]: Thank you so much for this

File: toolsrc/src/vcpkg/metrics.cpp [hunk 5]
```diff
@@ -248,6 +248,9 @@ namespace vcpkg::Metrics
         ;
     static bool g_should_print_metrics = false;
 
+#if defined(_MSC_VER)
+#pragma warning(suppress : 6326) // Potential comparison of a constant with another constant.
+#endif // _MSC_VER
     bool get_compiled_metrics_enabled() { return VCPKG_DISABLE_METRICS == 0; }
```
Comment (strega-nil) [inline toolsrc/src/vcpkg/metrics.cpp:254]: Maybe this would be better as `return !VCPKG_DISABLE_METRICS;`?

**Review by strega-nil — APPROVED** [review]
Minor suggestions, but this looks good :)

File: toolsrc/src/vcpkg/base/system.process.cpp (hunk 3 above)
Comment (BillyONeal) [reply toolsrc/src/vcpkg/base/system.process.cpp:389]: That would require the ability to emit the pragma in a macro and I'm unaware of reliable ways to do that.

File: toolsrc/src/vcpkg/metrics.cpp (hunk 5 above)
Comment (BillyONeal) [reply toolsrc/src/vcpkg/metrics.cpp:254]: Good idea!

File: toolsrc/src/vcpkg/base/system.process.cpp [hunk 6]
```diff
@@ -327,29 +327,47 @@ namespace vcpkg
...
+
+        void swap(ProcessInfo& other) noexcept
+        {
+            std::swap(proc_info.hProcess, other.proc_info.hProcess);
+            std::swap(proc_info.hThread, other.proc_info.hThread);
+        }
+
+        friend void swap(ProcessInfo& lhs, ProcessInfo& rhs) noexcept { lhs.swap(rhs); }
 
+        unsigned int wait_and_close_handles()
```
Comment (BillyONeal) [inline toolsrc/src/vcpkg/base/system.process.cpp:362]: This should just be called wait now

File: toolsrc/src/vcpkg/base/system.process.cpp (hunk 3 above)
Comment (strega-nil) [reply toolsrc/src/vcpkg/base/system.process.cpp:389]: `_Pragma`?

**Comment by strega-nil** [comment]
@BillyONeal PR for change ideas at BillyONeal#1

**Comment by BillyONeal** [comment]
@strega-nil Merged; also feel free to just push into it if you want :)

**Comment by BillyONeal** [comment]
Interestingly one of the failing tests says "// TODO: order here may change but A < X, and B anywhere" -- maybe we are seeing differences in std::sort between the standard libraries here?

**Comment by BillyONeal** [comment]
Map fixed it (but Linux is on the floor atm)

**Pull request merged**